### DISCUSSION TITLE
Regarding return values

### DISCUSSION THREAD

**Discussion by c0nwild** [elheck/Squiddrone#326] [Coding] [answered]
Hi all,
here is my problem:

In the COM driver architecture there is a basic layer that accesses HAL, an intermediate layer that provides the nrf24l01 specific command features (like write to register, read from register, write payload data, etc...) and finally a top layer that provides only "put data packet" and "get data packet" functionality as interface to the application layer (see https://github.com/Squiddrone/Squiddrone/wiki/UML-class-COM).

Now some private functions from the top layer (Class NRF24L01) kind of orchestrate other private functions to i.e. prepare the chip for tx or rx mode. Each function called has a return value of type DriverStatus to show if everything is happy. How do we handle a series of function calls where every return value might be a reason to throw an error at once? 
For example:

```
auto NRF24L01::InitTx() noexcept -> types::DriverStatus {
  auto rv;
  spi_protocol_->FlushTxBuffer();
  spi_protocol_->FlushRxBuffer();
  spi_protocol_->ReadAndClearIRQFlags().first;

  if (current_operation_mode_ == OperationMode::PRIM_TX) {
    return types::DriverStatus::OK;
  }

 SetOperationMode(OperationMode::PRIM_TX);
  /* needed in final implementation
    SetTxAddress(tx_addr);
    SetRxAddress(DataPipe::data_pipe_0, tx_addr);
  */

  EnableDataPipe(DataPipe::RX_PIPE_0);
  //EnableAutoAck(DataPipe::RX_PIPE_0);
  ConfigAutoRetransmission(AutoRetransmissionDelay::ARD500US, AutoRetransmitCount::ARC0);

  return rv;
}
```

Basically every function call has a return value that could indicate a HAL problem, timeout, etc...

**Comment by elheck** [comment, marked as answer]
Usually in desktop modern c++ you would use exceptions for this, because we need something that breaks the scope of a function call. As we are in an embedded environment, a error handling macro would suffice, because this can inject the return statement at every place. This is also common practice in most embedded environments.
